### PR TITLE
[Mime] Remove @internal from Headers methods

### PR DESCRIPTION
**Pull request by VincentLanglet**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #... 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

I don't understand why the methods
- `Headers::getHeaderBody()`
- `Headers::setHeaderBody()`
- `Headers::getHeaderParameter()`
- `Headers::setHeaderParameter()`
are marked as internal.

They are already used by others libraries, like 
https://github.com/symfony/mailchimp-mailer/blob/a6c3fa9aeaf83ad79445c9e7c072e15288bea3c9/Transport/MandrillApiTransport.php#L92
https://github.com/symfony/mailchimp-mailer/blob/a6c3fa9aeaf83ad79445c9e7c072e15288bea3c9/Transport/MandrillApiTransport.php#L99

The implementation shouldn't change so they could be under the BC promise.
Plus they are useful, and I don't see any other way that reimplementing the method if we want to do something like
```
->getHeaderParameter('Content-Disposition', 'name')
```

### CODE REVIEW

**Comment by Nyholm** [comment]
Could you elaborate why you want them to be not internal? What is your scenario where these methods are the only solution?

**Comment by VincentLanglet** [comment]
Sure, 

I followed the recommendation https://github.com/symfony/symfony/issues/40056#issuecomment-770936827 and implemented my own mailchimp mailer with more configuration options.
It implied that I have some code similar with the official symfony/mailchimp-mailer.

When I create the payload, it requires at least the same data than the one symfony is building in https://github.com/symfony/mailchimp-mailer/blob/a6c3fa9aeaf83ad79445c9e7c072e15288bea3c9/Transport/MandrillApiTransport.php#L73
And in this method, two internals method are used by symfony
https://github.com/symfony/mailchimp-mailer/blob/a6c3fa9aeaf83ad79445c9e7c072e15288bea3c9/Transport/MandrillApiTransport.php#L92-L99

If I want to reproduce the same behavior, I don't have any other options to get the `Content-Disposition`.

**Comment by Nyholm** [comment]
Thank you. I appreciate the context. 

I don't think this is a bug fix. It is a feature that we are making methods non-internal. Ie we should target 5.x. 

I'll wait for a review from someone with better understanding of the Mime component.

**Comment by OskarStark** [comment]
I agree, but I would go to merge this as a bugfix, because it just removes an annotation and opens sth instead of closing.

👍 for a bugfix

**Review by OskarStark — APPROVED** [review]

**Comment by Nyholm** [comment]
I consider this the same as making a private function public. Ie, we add a new function to the public API, hence a feature. 

But Im not sure how we usually do in scenarios like this.

**Comment by VincentLanglet** [comment]
The code in the 4.4 branch and the 5.x one for these internal methods are the same, so it won't hurt to merge it on 4.4.
It can fix some static analysis issues, like psalm reporting the usage of an internal method.

But I understand if you prefer to merge this on 5.x, I'll just ignore the error until the next version. Not a big deal.

**Comment by OskarStark** [comment]
friendly ping @nicolas-grekas for a decision 😍

**Comment by nicolas-grekas** [comment]
That's a new feature for sure: it will deserve dedicated changlog line + tests + doc.

**Comment by ro0NL** [comment]
IIUC the intended public API is `has()` and `get()`

edit: oh Headers has zero inheritance :+1: i agree this API is less boring.

**Comment by stof** [comment]
For instance, the `setHeaderBody` method does not properly validate the value for `$type` before calling the method with the name built dynamically based on the type. That's not really suitable as a non-internal method (for an internal method, it can be considered our own responsibility to use it only with valid values for the type, so skipping validation).

**Comment by VincentLanglet** [comment]
> For instance, the `setHeaderBody` method does not properly validate the value for `$type` before calling the method with the name built dynamically based on the type. That's not really suitable as a non-internal method (for an internal method, it can be considered our own responsibility to use it only with valid values for the type, so skipping validation).

What about keeping the setter as internal, but not the getters ?

**Comment by VincentLanglet** [comment]
PR was updated to only remove @internal on the getters.

File: UPGRADE-5.3.md
Comment (xabbuh) [inline UPGRADE-5.3.md]: [...] from the [...]

Comment (VincentLanglet) [reply UPGRADE-5.3.md]: Done :)

**Review by Nyholm — APPROVED** [review]

File: UPGRADE-5.3.md
Comment (xabbuh) [reply UPGRADE-5.3.md]: there is a second to that should be replaced with from :)

Comment (xabbuh) [reply UPGRADE-5.3.md]: Or maybe reword it like this:

> Remove the internal annotation from the `getHeaderBody()` and `getHeaderParameter()` methods of the `Headers` class.

**Review by OskarStark — APPROVED** [review]

File: UPGRADE-5.3.md
Comment (VincentLanglet) [reply UPGRADE-5.3.md]: Changed

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @VincentLanglet.

**Pull request merged**
